### DISCUSSION TITLE
Turn off all backends except snippets in Tab stop fields

### DISCUSSION THREAD

**Discussion by rickswe** [company-mode/company-mode#1377] [Q&A] [answered]
Hello, when doing a completion I only want my snippets coming up. How do I disable other backends
that are not yasnippet?

Setting it with `setq` works but then I will disable it globally, which I don't want. I'm not sure which
completions these are? I'm in emacs-lisp-mode.

I tried this in my snippet:

```snippet
# -*- mode: snippet -*-
# name: sub 
# key: sub 
# type: command
# --
(let ((company-backends '((:separate company-yasnippet ))))
  (yas-expand-snippet  "${1:Enter Subheading$$(when (and yas-modified-p yas-moving-away-p )(my/s-center yas-text :len \"large\" :beg \"[\" :end \"]\" ))}"))
```
![](https://i.imgur.com/suFSqL5.png)

Here is my company setup if needed:
```lisp
(setq company-tooltip-align-annotations t ; aligns annotation to the right
        company-tooltip-limit 12 ; bigger popup window
        company-idle-delay 0   ; decrease delay before autocompletion popup shows
        company-echo-delay (if (display-graphic-p) nil 0)
        company-minimum-prefix-length 2
        company-require-match nil
        company-dabbrev-ignore-case nil
        company-dabbrev-downcase nil
        company-dabbrev-minimum-length 10 ; Use buffer completion at 10 charcters.
        company-global-modes '(not erc-mode message-mode help-mode
                                   gud-mode eshell-mode shell-mode)
        ;;company-backends '((company-capf :with company-yasnippet)
        ;;                   (company-dabbrev-code company-keywords company-files)
        ;;                   company-dabbrev))

        ;; This puts snippet completion first.
        company-backends '((:separate company-yasnippet company-files company-bbdb company-semantic company-cmake company-capf company-clang
        company-dabbrev-code company-gtags company-keywords
        company-oddmuse company-dabbrev)))
```

I noticed that one suggestion is `:inside-paren-newline-start-from-block` is from `python-indent-context`
which I"m not using at all in my elisp code.

Thanks

**Comment by dgutov** [comment, marked as answer]
Using let-binding as in your example doesn't work because `yas-expand-snippet` exits right after it inserts the template with all the necessary structures into the buffer, so that snippet expansion works, but before you're able continue typing, initiate completion, etc.

To exclude all completions except yasnippet's, you could set `company-backends` to `'(company-yasnippet)` (a simpler value than in your example, though yours should also work). Maybe do it buffer-locally (see `setq-local`), to avoid affecting other buffers. Depending on whether the value is buffer-local already, you could back up the previous value to some new var, to be able to restore it later.

To do this alteration temporarily you could use yasnippet's hooks, such as `yas-before-expand-snippet-hook` and `yas-after-exit-snippet-hook`. In one hook function you would change the value, and in the other you would restore it. Maybe call `kill-local-variable` if it wasn't previously set locally.

I'm not quite sure why you would limit completion to snippets (I would rarely invoke snippet expansion inside another snippet), but the above is how you could do it.

**Comment by rickswe** [comment]
Thanks, that worked. Here is the code.

```lisp
(add-hook 'yas-before-expand-snippet-hook (lambda () (setq-local company-backends '(company-yasnippet))))
(add-hook 'yas-after-exit-snippet-hook    (lambda () (kill-local-variable 'company-backends)))
```

>I'm not quite sure why you would limit completion to snippets 

Because I'm trying to type something in the tab stop fields like `code` and something random  comes up
like `:ccl-decoder`. If I press tab now I completed `:ccl-decoder` instead of `code`. I'll need to figure out
later how to turn off these random suggestions. I think it may have something to do with one of the
`company-backends`. I don't think they are elisp or belong in `emacs-lisp-mode`.

>I would rarely invoke snippet expansion inside another snippet.

The reason for that is because I have snippets that create different sizes of a comment box. The snippets
subs,subm,subl etc... call the function `my/s-center` with different arguments. If I didn't call `my/s-center`
I would have to duplicate the code across multiple snippets.

**Reply by dgutov** [reply]
Symbols starting with `:` are called "keyword symbols", they evaluate to themselves. And as soon as such a symbol is used, it's added to the session's obarray. Emacs Lisp completion works off the contents of obarray. Having completions like `:Kind`, etc, is valid.

Why do they match "in"? Apparently, you have set up some permissive completion style that allows substring completion. `:K`*in*`d`. E.g. I can get the same effect with `(setq completion-styles '(flex))`.

To get closer to the default behavior, try using some simpler completion styles in front of it, like: `(setq completion-styles '(basic partial-completion flex))`. That exhibits a more subtle behavior, though, where you won't see the more lax completions while you still have either `basic` or `partial-completion` matches for your current input (the first style to provide completions wins).

There is a different way to get rid on unwanted completions, however: add a filtering function to `company-transformers`.

**Comment by rickswe** [comment]
>Symbols starting with : are called "keyword symbols", they evaluate to themselves.

The suggestions would be useful if I was writing function that took those keys like most `cl` functions.
If I'm typing `(cl-member ` then I would expect `:test` to come up when I type `:`. Not sure if
I'm asking too much or if that's possible.

The problem was I had `orderless` first in completion-styles, doing your suggestion worked.

Thanks

**Reply by dgutov** [reply]
It's not just `cl` functions that use keyword symbols as arguments. `add-function`, for example, also uses them. And some programmer might be looking to manipulate the value. But indeed they are in the minority, and most of the time you don't use them.

So filtering them out in a "smart" way is hard. Anyway, that would be a job for `elisp-completion-at-point` (function inside Emacs core), so feel free to file a bug. Maybe someone will have better ideas for improvement.

Happy to help.

**Comment by rickswe** [comment]
@dgutov  I have a similar issue. I'm trying to setup react on Emacs but the problem is I have to
use "C-." to turn off completion ( don't remember where this is mapped) then followed
by "C-d" to insert the tag. I would like to do "C-d" to insert the tag but it starts deleting text,
perhaps this is Emacs default behavior. The plan would be to turn off completion in the lambda?
Not sure if this is possible.



```lisp
(setq evil-want-C-d-scroll nil) ; rjsx-mode-map
(use-package rjsx-mode
  :straight t
  :config
  (evil-define-key 'insert rjsx-mode-map (kbd "C-d") #'(lambda ()(interactive)
(let ((company-backends '((:separate company-yasnippet ))))
(call-interactively 'rjsx-delete-creates-full-tag)
  )))
  :mode ("\\.js\\'" . rjsx-mode))
```

Thanks in advance.

In this video:
1. C-. to turn off completion then C-d to complete the tag
2. C-d - deletes text, not what I want.

https://github.com/company-mode/company-mode/assets/76216665/99dade07-d6af-4d1e-b83b-0fc8bb3f6594

**Reply by dgutov** [reply]
This looks like a totally different question (no relation to yasnippet).

Anyway, I'm not sure why this is not working for you. In plain Emacs configuration, the new `C-d` binding does what your binding specifies here. Perhaps Evil overrides the binding and does not restore it entirely (so that the major mode bindings could be followed).

You can try binding it in `company-active-map` instead. Just keep in mind that the result is it will be available in all major modes, not just rjsx.
